utils/sync/plocker.go

```diff
@@ -0,0 +1,46 @@
+package sync
+
+import "sync"
+
+// PartitionLocker is a lock that can be used to lock different partitions at the same time.
+type PartitionLocker struct {
+	c *sync.Cond
+	l sync.Locker
+	s map[string]struct{}
+}
+
+// NewPartitionLocker returns a new PartitionLocker.
+func NewPartitionLocker() *PartitionLocker {
+	l := &sync.Mutex{}
+	return &PartitionLocker{
+		c: sync.NewCond(l),
+		l: l,
+		s: make(map[string]struct{}),
+	}
+}
+
+func (p *PartitionLocker) locked(id string) (ok bool) {
+	_, ok = p.s[id]
+	return
+}
+
+// Lock locks the lock. If the lock is locked, it waits until the lock is unlocked.
+func (p *PartitionLocker) Lock(id string) {
+	p.l.Lock()
+	defer p.l.Unlock()
+	for p.locked(id) {
+		p.c.Wait()
+	}
+	p.s[id] = struct{}{}
+}
```

Comment:
Assume the following use case:
```		
		go func() {
				locker.Lock("id1")
                 time.Sleep(1 * time.Millisecond)
				locker.Unlock("id1")
         }()
         go func() {
				locker.Lock("id2")
                 time.Sleep(1 * time.Minute)
				locker.Unlock("id2")
         }()
         go func() {
				locker.Lock("id2")
                 time.Sleep(1 * time.Millisecond)
				locker.Unlock("id2")
         }()
 ```
 Depending on the execution order, id1 can wait unnecessarily long for id2 to get locked.
 Since both will try to hold: `p.l.Lock()`
 
 Depending on your use case, this might or might not be a problem.

Reply:
`p.l` is not locked while `p.c.Wait` is blocking. Thus I can't see where the problem can be